### PR TITLE
add inference example for Deep Speech 2

### PR DESCRIPTION
**Pull request by hhbyyh**
resolve some building issue and add inference example.

Upgrading the example to BigDL 0.3 is a follow-up task.

### CODE REVIEW

File: pipeline/deepspeech2/pom.xml [hunk 1]
```diff
@@ -77,22 +77,21 @@
...
-            </dependency>
-            <dependency>
-                <groupId>com.github.fommil</groupId>
-                <artifactId>jniloader</artifactId>
-                <version>1.1</version>
-            </dependency>
+        <dependency>
+            <groupId>com.intel.analytics.bigdl</groupId>
+            <artifactId>bigdl-SPARK_2.0</artifactId>
+            <version>0.1.0</version>
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/pom.xml]: why not 0.3.0-SNAPSHOT?

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+@SerialVersionUID( - 467695939363389565L)
+class BatchNormalizationDS[@specialized(Float, Double) T: ClassTag](
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala]: please add some java doc

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  nOutput: Int,
+  eps: Double = 1e-3,
+  momentum: Double = 0.1,
+  affine: Boolean = false)
+  (implicit ev: TensorNumeric[T])
+  extends BatchNormalization[T](nOutput, eps, momentum, affine) {
+
+  val batchDim = 0
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala]: I think batchDim should be 1, to keep same meaning with Recurrent and BiRecurrent, etc.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  val timeDim = 1
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala]: same

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  val featDim = 2
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim == 3,
+      "In BatchNormalizationBRNN, input should be a 3D tensor, [B, T, D]" +
+        s"input.dim = ${input.dim}")
+    val size = input.size()
+    input.resize(size(batchDim) * size(timeDim), size(featDim))
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala]: you can create an array to store (size(batchDim) * size(timeDim), size(featDim)), else "resize" will create array each time.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BiRecurrentDS.scala [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+@SerialVersionUID( - 643193217505024792L)
+class BiRecurrentDS[T : ClassTag](
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BiRecurrentDS.scala]: add some docs, and what's the difference with BiRecurrent?

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BifurcateSplitTable.scala [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    val slices = input.size(dimension)
+    require(slices >= 1,
+      s"BifurcateSplitTable: the size of referred dimension is ${slices}. " +
+        s"It should be larger than 1.")
+    val leftSlices = slices >> 1
+    val rightSlices = slices - leftSlices
+
+    output(1) = input.narrow(dimension, 1, leftSlices).contiguous()
+    output(2) = input.narrow(dimension, 1 + leftSlices, rightSlices).contiguous()
+    output
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BifurcateSplitTable.scala]: output may use same memory with input. It is not recommended.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BifurcateSplitTable.scala [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+ * the specified dimension `dimension`.
+ *
+ * The input to this layer is expected to be a tensor, or a batch of tensors;
+ *
+ * @param dimension to be split along this dimension
+ * @tparam T Numeric type. Only support float/double now
+ */
+
+@SerialVersionUID( 6506746847756687944L)
+class BifurcateSplitTable[T: ClassTag](
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BifurcateSplitTable.scala]: same name with BifurcateSplitTable in bigdl?

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/pipeline/acoustic/DeepSpeech2ModelLoader.scala [hunk 9]
```diff
@@ -65,11 +65,11 @@ class DeepSpeech2ModelLoader[T : ClassTag](depth: Int = 1)
   : Module[T] = {
     val isSplitInput = if (curDepth == 1) false else true
     val layers = Sequential()
-        .add(BiRecurrent[T](
-          JoinTable[T](2, 2),
-          batchNormParams = BatchNormParams[T](eps = 0.001, affine = false),
-          isSplitInput = isSplitInput)
-          .add(RnnCell[T](inputSize, hiddenSize, HardTanh[T](0, 20, true), isInputWithBias = false)))
+//        .add(BiRecurrent[T](
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/pipeline/acoustic/DeepSpeech2ModelLoader.scala]: why delete?

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/example/InferenceExample.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.spark.ml.feature.FlacReader
+import org.apache.spark.sql.SparkSession
+
+object InferenceExample {
+
+  Logger.getLogger("org").setLevel(Level.WARN)
+  val logger = Logger.getLogger(getClass)
+
+  def main(args: Array[String]): Unit = {
+    parser.parser.parse(args, LocalOptimizerPerfParam()).foreach { param =>
```
Comment (zhangxiaoli73) [inline pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/example/InferenceExample.scala]: name "LocalOptimizerPerfParam" is not good. This example has nothing to do with LocalOptimizerPerf

**Comment by hhbyyh** [comment]
Thanks @zhangxiaoli73 . As in the PR description, the PR focus on providing the inference example based on BigDL 0.1.

There will be a follow up PR for upgrading the code to BigDL 0.3. And all the non-example code will be removed.

**Comment by hhbyyh** [comment]
I'll merge this to unblock customer first.

File: pipeline/deepspeech2/pom.xml (hunk 1 above)
Comment (hhbyyh) [reply pipeline/deepspeech2/pom.xml]: the code was originally developed with BigDL 0.1.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala (hunk 2 above)
Comment (hhbyyh) [reply pipeline/deepspeech2/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalizationDS.scala]: We'll get rid of the duplicate code after upgrading to 0.3.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/pipeline/acoustic/DeepSpeech2ModelLoader.scala (hunk 9 above)
Comment (hhbyyh) [reply pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/pipeline/acoustic/DeepSpeech2ModelLoader.scala]: all training code should be removed.

File: pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/example/InferenceExample.scala (hunk 10 above)
Comment (hhbyyh) [reply pipeline/deepspeech2/src/main/scala/com/intel/analytics/zoo/pipeline/deepspeech2/example/InferenceExample.scala]: Let's change this when upgrading.

**Comment by hhbyyh** [comment]
Still need green light from reviewer @zhangxiaoli73

**Comment by hhbyyh** [comment]
Please understand this is important to unblock the customer. @zhangxiaoli73 @jason-dai

**Comment by zhangxiaoli73** [comment]
example test good
